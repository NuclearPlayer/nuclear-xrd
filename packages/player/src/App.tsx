import { createRouter, RouterProvider } from '@tanstack/react-router';
import { MotionConfig } from 'framer-motion';

import { routeTree } from './routeTree.gen';

const router = createRouter({ routeTree });

declare module '@tanstack/react-router' {
  interface Register {
    router: typeof router;
  }
}

function App() {
  return (
    <MotionConfig
      reducedMotion={process.env.NODE_ENV === 'test' ? 'always' : 'user'}
    >
      <RouterProvider router={router} />
    </MotionConfig>
  );
}

export default App;
