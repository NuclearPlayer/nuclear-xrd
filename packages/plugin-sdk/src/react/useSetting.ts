import { useEffect, useMemo, useRef, useState } from 'react';

import type { SettingsHost, SettingValue } from '../types/settings';

export const useSetting = <T extends SettingValue = SettingValue>(
  host: SettingsHost | undefined,
  id: string,
) => {
  const hostRef = useRef(host);
  hostRef.current = host;

  const [currentValue, setCurrentValue] = useState<T | undefined>(undefined);

  useEffect(() => {
    if (!hostRef.current) return;
    let unsubscribe: (() => void) | undefined;
    let isMounted = true;

    hostRef.current.get<T>(id).then((initialValue) => {
      if (!isMounted) return;
      setCurrentValue(initialValue);
      unsubscribe = hostRef.current?.subscribe<T>(id, (nextValue) =>
        setCurrentValue(nextValue),
      );
    });

    return () => {
      isMounted = false;
      if (unsubscribe) unsubscribe();
    };
  }, [id, hostRef]);

  const setValue = useMemo(
    () => (nextValue: T) => {
      if (!hostRef.current) return;
      void hostRef.current.set<T>(id, nextValue);
    },
    [id],
  );

  return [currentValue, setValue] as const;
};
