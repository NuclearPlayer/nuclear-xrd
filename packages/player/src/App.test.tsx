import { render } from '@testing-library/react';

import App from './App';
import { AppWrapper } from './App.test-wrapper';

describe('App', () => {
  beforeEach(() => {
    AppWrapper.resetState();
  });

  it('should render snapshot', () => {
    const component = render(<App />);
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should handle sidebar collapse/expand user flow', async () => {
    render(<App />);

    expect(AppWrapper.getLayoutState().leftSidebar.isCollapsed).toBe(false);
    expect(AppWrapper.getLayoutState().rightSidebar.isCollapsed).toBe(false);

    await AppWrapper.toggleLeftSidebar();
    expect(AppWrapper.getLayoutState().leftSidebar.isCollapsed).toBe(true);
    expect(AppWrapper.getLayoutState().rightSidebar.isCollapsed).toBe(false);

    await AppWrapper.toggleRightSidebar();
    expect(AppWrapper.getLayoutState().leftSidebar.isCollapsed).toBe(true);
    expect(AppWrapper.getLayoutState().rightSidebar.isCollapsed).toBe(true);

    await AppWrapper.toggleLeftSidebar();
    expect(AppWrapper.getLayoutState().leftSidebar.isCollapsed).toBe(false);
    expect(AppWrapper.getLayoutState().rightSidebar.isCollapsed).toBe(true);

    await AppWrapper.toggleRightSidebar();
    expect(AppWrapper.getLayoutState().leftSidebar.isCollapsed).toBe(false);
    expect(AppWrapper.getLayoutState().rightSidebar.isCollapsed).toBe(false);

    await AppWrapper.toggleLeftSidebar();
    await AppWrapper.toggleLeftSidebar();
    expect(AppWrapper.getLayoutState().leftSidebar.isCollapsed).toBe(false);
  });
});
