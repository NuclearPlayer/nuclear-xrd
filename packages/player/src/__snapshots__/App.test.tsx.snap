// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > should render snapshot 1`] = `
<DocumentFragment>
  <div
    class="h-screen w-screen grid grid-rows-[auto_1fr_auto] "
  >
    <header
      class="h-12 bg-background border-b-2 border-border flex items-center px-4 "
    />
    <div
      class="grid grid-cols-[auto_1fr_auto] h-full bg-background-secondary "
    >
      <div
        class="bg-background-secondary border-border flex flex-col relative p-2 border-r-2"
      >
        <span
          class="flex flex-row items-center mb-4 justify-end"
        >
          <button
            class="inline-flex items-center transition-all rounded text-foreground bg-primary border-2 border-border shadow-shadow hover:translate-x-shadow-x hover:translate-y-shadow-y hover:shadow-none size-10 top-2 px-2 right-1"
          >
            <svg
              class="lucide lucide-panel-left"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="18"
                rx="2"
                width="18"
                x="3"
                y="3"
              />
              <path
                d="M9 3v18"
              />
            </svg>
          </button>
        </span>
        <div
          class="flex-1 overflow-hidden"
          style="opacity: 0; transform: translateX(-20px);"
        >
          <div
            class="flex flex-col justify-start items-start flex-1"
          >
            <div
              class="flex flex-col gap-2 w-full"
            >
              <button
                class="transition-all rounded text-foreground bg-transparent inline-flex flex-row items-center justify-between"
              >
                Navigation
                <div>
                  <svg
                    class="lucide lucide-chevron-right"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m9 18 6-6-6-6"
                    />
                  </svg>
                </div>
              </button>
            </div>
            <div
              class="flex flex-col gap-2 w-full"
            >
              <button
                class="transition-all rounded text-foreground bg-transparent inline-flex flex-row items-center justify-between"
              >
                Your library
                <div>
                  <svg
                    class="lucide lucide-chevron-right"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m9 18 6-6-6-6"
                    />
                  </svg>
                </div>
              </button>
            </div>
          </div>
        </div>
        <div
          class="absolute top-0 bottom-0 w-1 cursor-col-resize hover:bg- transition-colors right-0"
        />
      </div>
      <main
        class="bg-background-secondary overflow-auto"
      />
      <div
        class="bg-background-secondary border-border flex flex-col relative p-2 border-l-2"
      >
        <span
          class="flex flex-row items-center mb-4 justify-start"
        >
          <button
            class="inline-flex items-center transition-all rounded text-foreground bg-primary border-2 border-border shadow-shadow hover:translate-x-shadow-x hover:translate-y-shadow-y hover:shadow-none size-10 top-2 px-2 left-1"
          >
            <svg
              class="lucide lucide-panel-right"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="18"
                rx="2"
                width="18"
                x="3"
                y="3"
              />
              <path
                d="M15 3v18"
              />
            </svg>
          </button>
        </span>
        <div
          class="flex-1 overflow-hidden"
          style="opacity: 0; transform: translateX(20px);"
        />
        <div
          class="absolute top-0 bottom-0 w-1 cursor-col-resize hover:bg- transition-colors left-0"
        />
      </div>
    </div>
    <footer
      class="h-16 bg-background border-t-2 border-border flex items-center px-4 "
    />
  </div>
</DocumentFragment>
`;
